lib/freebl/algcmac.c

```diff
@@ -0,0 +1,327 @@
+// cipher.
+typedef enum {
+    CMAC_AES = 0
+} CMACCiphers;
+
+struct CMACContextStr {
+    // Information about the block cipher to use internally. The cipher should
+    // be placed in ECB mode so that we can use it to directly encrypt blocks.
+    // Store a pointer to the cipher's context and keep track of the block
+    // size so we can validate input and return parameters ourselves.
+    CMACCiphers cipherType;
+    union {
+        void *raw;
+        AESContext *aes;
+    } cipher;
+    int blockSize;
+
+    // Whether or not we allocated (via CMAC_Create) this CMACContext. This
+    // affects CMAC_Destroy.
+    PRBool wasAllocated;
+
+    // Internal keys which are conditionally used by the algorithm. Derived
+    // from encrypting the NULL block. We leave the storing of (and the
+    // cleanup of) the CMAC key to the underlying block cipher.
+    unsigned char *k1;
+    unsigned char *k2;
+
+    // When Update is called with data which isn't a multiple of the block
+    // size, we need a place to put it. HMAC handles this by passing it to
+    // the underlying hash function right away; we can't do that as the
+    // contract on the cipher object is different.
+    unsigned int partialIndex;
+    unsigned char *partialBlock;
+
+    // Last encrypted block. This gets xor-ed with partialBlock prior to
+    // encrypting it. NIST defines this to be the empty string to begin.
+    unsigned char *lastBlock;
+};
+
+/* byte arrays are really the wrong level for this: we want to shift the
+ * entire block left by one bit. With AES, this ends up being a 128-bit
+ * block size that we have to deal with manually; a union data structure
+ * between a __uint128_t and a byte array would be better. */
+static unsigned char *left_shift(unsigned char *data, int length) {
+    unsigned char *result = PORT_ZNewArray(unsigned char, length);
+
+    for (int i = 0; i < length; i++) {
+        result[i] = data[i] << 1;
+        if (i + 1 < length) {
+            result[i] |= data[i+1] >> 7;
+        }
+    }
+
+    return result;
+}
+
+static SECStatus cmac_Encrypt(CMACContext *ctx, unsigned char *output,
+                              unsigned int *outputLen,
+                              unsigned int maxOutputLen,
+                              const unsigned char *input,
+                              unsigned int inputLen) {
+    if (ctx->cipherType == CMAC_AES) {
+        unsigned int tmpOutputLen;
+        SECStatus ret = AES_Encrypt(ctx->cipher.aes, output, &tmpOutputLen, maxOutputLen, input, inputLen);
+
+        if (outputLen != NULL) {
+            *outputLen = tmpOutputLen;
+        }
+
+        return ret;
+    }
+
+    return SECFailure;
+}
+
+/* NIST SP.800-38B, 6.1 Subkey Generation */
+static SECStatus cmac_GenerateSubkeys(CMACContext *ctx) {
+    unsigned char *null_block = PORT_ZNewArray(unsigned char, ctx->blockSize);
+    unsigned char *L = PORT_ZNewArray(unsigned char, ctx->blockSize);
```

Comment:
You could make these stack variables with size MAX_BLOCK_SIZE. and same the Alloc.

Reply:
ACK.